### PR TITLE
Fix unused gradient tracking to respect create_graph

### LINKED ISSUE #168059: Gradients wrt unused tensors are still marked with requires_grad=True if allow_unused=True
### 🐛 Describe the bug

When taking gradients with ``torch.autograd.grad`` and ``allow_unused=True``, the resulting gradient sometimes has ``requires_grad=True`` and sometimes does not, depending on whether the input tensor was actually used in the graph. This is despite create and retain graph both being set to ``False``:

```python
x = torch.ones(10).requires_grad_(True)
y = torch.ones(10).requires_grad_(True)
z = (x**2).sum()
g = torch.autograd.grad(z, (x, y), allow_unused=True, materialize_grads=True, retain_graph=False, create_graph=False)

g[0].requires_grad # False, expected
g[1].requires_grad # True, unexpected
```

### Versions

```PyTorch version: 2.9.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.11 | packaged by Anaconda, Inc. | (main, Jun  5 2025, 13:09:17) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-126-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.64
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA H200
GPU 1: NVIDIA H200
GPU 2: NVIDIA H200
GPU 3: NVIDIA H200
GPU 4: NVIDIA H200
GPU 5: NVIDIA H200
GPU 6: NVIDIA H200
GPU 7: NVIDIA H200

Nvidia driver version: 570.158.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               160
On-line CPU(s) list:                  0-159
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8460Y+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   40
Socket(s):                            2
Stepping:                             8
Frequency boost:                      enabled
CPU max MHz:                          2001.0000
CPU min MHz:                          800.0000
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            3.8 MiB (80 instances)
L1i cache:                            2.5 MiB (80 instances)
L2 cache:                             160 MiB (80 instances)
L3 cache:                             210 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-39,80-119
NUMA node1 CPU(s):                    40-79,120-159
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mkl_fft==1.3.11
[pip3] mkl_random==1.2.8
[pip3] mkl-service==2.4.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] optree==0.17.0
[pip3] torch==2.9.1
[pip3] torchopt==0.7.3
[pip3] torchviz==0.0.3
[pip3] triton==3.5.1
[conda] blas                         1.0              mkl
[conda] cudatoolkit                  11.7.0           hd8887f6_10      nvidia
[conda] cudatoolkit-dev              11.7.0           h1de0b5d_6       conda-forge
[conda] intel-openmp                 2025.0.0         h06a4308_1171
[conda] mkl                          2025.0.0         hacee8c2_941
[conda] mkl-service                  2.4.0            py312h5eee18b_3
[conda] mkl_fft                      1.3.11           py312hacdc0fc_1
[conda] mkl_random                   1.2.8            py312h2fd27a0_1
[conda] numpy                        2.3.2            pypi_0           pypi
[conda] numpy-base                   1.26.4           py312he1678cf_1
[conda] nvidia-cublas-cu12           12.8.4.1         pypi_0           pypi
[conda] nvidia-cuda-cupti-cu12       12.8.90          pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu12       12.8.93          pypi_0           pypi
[conda] nvidia-cuda-runtime-cu12     12.8.90          pypi_0           pypi
[conda] nvidia-cudnn-cu12            9.10.2.21        pypi_0           pypi
[conda] nvidia-cufft-cu12            11.3.3.83        pypi_0           pypi
[conda] nvidia-curand-cu12           10.3.9.90        pypi_0           pypi
[conda] nvidia-cusolver-cu12         11.7.3.90        pypi_0           pypi
[conda] nvidia-cusparse-cu12         12.5.8.93        pypi_0           pypi
[conda] nvidia-cusparselt-cu12       0.7.1            pypi_0           pypi
[conda] nvidia-nccl-cu12             2.27.5           pypi_0           pypi
[conda] nvidia-nvjitlink-cu12        12.8.93          pypi_0           pypi
[conda] nvidia-nvtx-cu12             12.8.90          pypi_0           pypi
[conda] optree                       0.17.0           pypi_0           pypi
[conda] tbb                          2022.0.0         hdb19cb5_0
[conda] tbb-devel                    2022.0.0         hdb19cb5_0
[conda] torch                        2.9.1            pypi_0           pypi
[conda] torchopt                     0.7.3            pypi_0           pypi
[conda] torchviz                     0.0.3            pypi_0           pypi
[conda] triton                       3.5.1            pypi_0           pypi```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan

### PR DESCRIPTION
**Pull request by dsashidh**
Fixes https://github.com/pytorch/pytorch/issues/168059

PyTorch was incorrectly setting requires_grad=True on unused gradients even when create_graph=False. This caused unnecessary autograd tracking and extra memory usage.

This PR updates the logic to set requires_grad based on the value of create_graph, ensuring that unused gradients are tracked only when explicitly requested. This aligns torch.autograd.grad( ) behavior with expectations.

Includes test coverage in test_unused_grad_requires_grad_with_materialize

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]
Thanks!

Technically changes behavior, if someone relied on gradients being tracked here it would no longer be tracked. Should be fine though as this is a bug fix.
